### PR TITLE
Base controllers02 - pagination+sorting

### PR DESCRIPTION
**Pull request by HerbertBodner**
async methods, CountAsync and GetAllAsync separated in Repository
SortBySplitter including unit tests

@yshchohaleu , @rurku the discussed implementation for sorting using expressions is not done yet. Feel free to take that challenge

### CODE REVIEW

File: src/WaCore.Crud.Utils/PagedList.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using System.Text;
+using WaCore.Crud.Contracts.Utils;
+
+namespace WaCore.Crud.Utils
+{
+    public class PagedList<T> : IPagedList<T>
+    {
+        private IList<T> _list;
+
+        public int TotalCount { get; private set; }
```
Comment (yshchohaleu) [inline src/WaCore.Crud.Utils/PagedList.cs]: Small issues - we don't need
```
private set;
```
here.

File: src/WaCore.Crud.Utils/PagedList.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            Offset = offset;
+            Limit = limit;
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            return _list.GetEnumerator();
+        }
+
+        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
```
Comment (yshchohaleu) [inline src/WaCore.Crud.Utils/PagedList.cs]: Please remove System.Collections namespace

File: src/WaCore.Crud.Utils/PagedList.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            return this.GetEnumerator();
```
Comment (yshchohaleu) [inline src/WaCore.Crud.Utils/PagedList.cs]: Please remove this.

File: src/WaCore.Crud.Utils/PagedList.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        public int TotalCount { get; private set; }
+
+        public int Offset { get; private set; }
+
+        public int? Limit { get; private set; }
+
+        public IList<T> List
+        {
+            get { return _list; }
```
Comment (yshchohaleu) [inline src/WaCore.Crud.Utils/PagedList.cs]: Let's use the latest features of C#
```
public IList<T> List => _list;
```

File: src/WaCore.Crud.Services/WacListDataService.cs [hunk 5]
```diff
@@ -1,23 +1,45 @@
...
 {
     public abstract class WacListDataService<TEntity, TFilter, TDto> : IWacListDataService<TFilter, TDto>
+        where TFilter : IWacFilter
     {
-        private readonly IWacListDataRepository<TEntity, TFilter> repo;
+        protected readonly IWacListDataRepository<TEntity, TFilter> repo;
 
         public WacListDataService(IWacListDataRepository<TEntity, TFilter> repo)
         {
             this.repo = repo;
```
Comment (yshchohaleu) [inline src/WaCore.Crud.Services/WacListDataService.cs]: Can you please rename it to Repo and remove this?

File: src/WaCore.Crud.UnitTests/Utils/SortBySplitterTests.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using WaCore.Crud.Utils;
+using Xunit;
```
[the remark this answers is not in the document]
Comment (HerbertBodner) [reply src/WaCore.Crud.UnitTests/Utils/SortBySplitterTests.cs]: yeah, that's done

File: src/WaCore.Crud.UnitTests/Utils/SortBySplitterTests.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        [Fact]
+        public void SplitSortByStringWhenTwoFieldsAndLeadingMinusReturnsSecondtCorrectOrderItem()
+        {
+            Validate("-myField+Field2", 2, OrderItem.OrderBy.Ascending, "Field2", 1);
+        }
+
+
+        private void Validate(string inputString, int expectedFieldCount, OrderItem.OrderBy expectedOrderBy, string expectedFieldName, int itemNrToValidate = 0)
```
[the remark this answers is not in the document]
Comment (HerbertBodner) [reply src/WaCore.Crud.UnitTests/Utils/SortBySplitterTests.cs]: ok, done

File: src/WaCore.Crud.Utils/PagedList.cs (hunk 1 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Utils/PagedList.cs]: sure, done :)

File: src/WaCore.Crud.Utils/PagedList.cs (hunk 4 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Utils/PagedList.cs]: yep

File: src/WaCore.Crud.Utils/PagedList.cs (hunk 2 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Utils/PagedList.cs]: as well :)

File: src/WaCore.Crud.Utils/PagedList.cs (hunk 3 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Utils/PagedList.cs]: ok, done

File: src/WaCore.Crud.Services/WacListDataService.cs (hunk 5 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Services/WacListDataService.cs]: yep, done

**Pull request merged**